### PR TITLE
Show class that originally defined attribute in output

### PR DESCRIPTION
**Pull request by ucodery**
If a subclass or its instance attempts to assign an incompatible type
display the class that originally defined the type.

Fixes issue #2500

### CODE REVIEW

File: mypy/checker.py [hunk 1]
```diff
@@ -2599,7 +2599,8 @@ def check_simple_assignment(self, lvalue_type: Optional[Type], rvalue: Expressio
             return rvalue_type
 
     def check_member_assignment(self, instance_type: Type, attribute_type: Type,
-                                rvalue: Expression, context: Context) -> Tuple[Type, Type, bool]:
+                                rvalue: Expression, context: MemberExpr
```
Comment (ilevkivskyi) [inline mypy/checker.py]: I would rename `context` to `lvalue`. Context is typically something that is used mostly for error line/column.

File: mypy/checker.py [hunk 2]
```diff
@@ -2612,9 +2613,17 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
         care about interaction between binder and __set__().
         """
         # Descriptors don't participate in class-attribute access
-        if ((isinstance(instance_type, FunctionLike) and instance_type.is_type_obj()) or
-                isinstance(instance_type, TypeType)):
-            rvalue_type = self.check_simple_assignment(attribute_type, rvalue, context)
+        if (isinstance(instance_type, CallableType) and instance_type.is_type_obj() and
+                isinstance(instance_type.ret_type, Instance)):
```
Comment (ilevkivskyi) [inline mypy/checker.py]: Named tuple class object have a `TupleType` return type, you can "convert" it to instance by using its fallback.

File: mypy/checker.py [hunk 3]
```diff
@@ -2674,6 +2684,52 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
         infer = is_subtype(rvalue_type, get_type) and is_subtype(get_type, set_type)
         return rvalue_type if infer else set_type, get_type, infer
 
+    def check_parent_member_assignment(self, lvalue_type: Type,
+                                       lvalue_attribute_type: Type, rvalue_type: Type,
+                                       rvalue: Expression, context: MemberExpr) -> Type:
```
Comment (ilevkivskyi) [inline mypy/checker.py]: Also `context` -> `lvalue` here.

File: mypy/checker.py [hunk 4]
```diff
@@ -2674,6 +2684,52 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
         infer = is_subtype(rvalue_type, get_type) and is_subtype(get_type, set_type)
         return rvalue_type if infer else set_type, get_type, infer
 
+    def check_parent_member_assignment(self, lvalue_type: Type,
+                                       lvalue_attribute_type: Type, rvalue_type: Type,
```
Comment (ilevkivskyi) [inline mypy/checker.py]: The naming here is also confusing. `lvalue_type` is actually `lvalue_base_type`, while `lvalue_attribute_type` is just `lvalue_type`.

File: mypy/checker.py [hunk 5]
```diff
@@ -2674,6 +2684,52 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
...
+                                       rvalue: Expression, context: MemberExpr) -> Type:
+        """Type inherited member assignment.
+
+        This defers to check_simple_assignment if the attribute is local to the instance_type,
+        otherwise it will inform the user where the attribute type was originally defined.
+
+        Return the inferred rvalue_type.
+        """
+        if isinstance(lvalue_type, (Instance)):
```
Comment (ilevkivskyi) [inline mypy/checker.py]: Parentheses are redundant here.

File: mypy/checker.py [hunk 6]
```diff
@@ -2674,6 +2684,52 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
...
+            attribute_is_defined = lvalue_type.has_readable_member(context.name)
+            if isinstance(lvalue_type, UnionType):
```
Comment (ilevkivskyi) [inline mypy/checker.py]: This branch is unreachable (see `if` just above). Also I think we can't give any better error message for unions, so I would just remove this branch.

File: mypy/checker.py [hunk 7]
```diff
@@ -2674,6 +2684,52 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
...
+                direct_lvalue_vars = list(itertools.chain.from_iterable(
+                                          i.type.names for i in lvalue_type.items))
+                lvalue_bases = list(itertools.chain.from_iterable(
+                                    i.type.bases for i in lvalue_type.items))
+            else:
+                direct_lvalue_vars = lvalue_type.type.names
+                lvalue_bases = lvalue_type.type.bases
+                if '__metaclass__' in lvalue_type.type.names:
+                    # if the assigned metaclass is not class itself we have other problems
+                    metacls = lvalue_type.type.names['__metaclass__'].type.ret_type  # type: ignore
```
Comment (ilevkivskyi) [inline mypy/checker.py]: This is a bed way of getting the metaclass. You should just use `metaclass_type` of the `TypeInfo` object.

File: mypy/checker.py [hunk 8]
```diff
@@ -2674,6 +2684,52 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
...
+                    lvalue_bases.append(metacls)
+                    attribute_is_defined = (attribute_is_defined or
+                                            metacls.has_readable_member(context.name))
+            if attribute_is_defined and context.name not in direct_lvalue_vars:
```
Comment (ilevkivskyi) [inline mypy/checker.py]: Why do you need to check for `attribute_is_defined`, to avoid this extra message if the type is from a `__getattr__()`? Please add a comment. Naively, if an attribute is not defined at all, we can't know its type.

File: mypy/checker.py [hunk 9]
```diff
@@ -2674,6 +2684,52 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
...
+                for base in lvalue_bases:
+                    if context.name in base.type.names:
+                        parent_context = base.type.names[context.name].node
+                        if isinstance(parent_context, Var) and parent_context.info.is_generic():
+                            # attribute may have been defined in this class but its type is
```
Comment (ilevkivskyi) [inline mypy/checker.py]:
```suggestion
                            # Attribute may have been defined in this class but its type is
```

File: mypy/checker.py [hunk 10]
```diff
@@ -2630,7 +2639,8 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
             # the return type of __get__. This doesn't match the python semantics,
             # (which allow you to override the descriptor with any value), but preserves
             # the type of accessing the attribute (even after the override).
-            rvalue_type = self.check_simple_assignment(get_type, rvalue, context)
+            rvalue_type = self.check_parent_member_assignment(instance_type, get_type,
+                                                              attribute_type, rvalue, context)
```
Comment (ilevkivskyi) [inline mypy/checker.py]: Why do you need to pass _both_ `attribute_type` and `get_type` here? This looks wrong, I think you should only pass `get_type`.

File: mypy/checker.py [hunk 11]
```diff
@@ -2674,6 +2684,52 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
...
+            if attribute_is_defined and context.name not in direct_lvalue_vars:
+                for base in lvalue_bases:
+                    if context.name in base.type.names:
+                        parent_context = base.type.names[context.name].node
+                        if isinstance(parent_context, Var) and parent_context.info.is_generic():
+                            # attribute may have been defined in this class but its type is
+                            # probably defined elsewhere, possibly in the instance itself.
+                            continue
+                        always_allow_any = (rvalue_type is not None and
+                                            not isinstance(rvalue_type, AnyType))
```
Comment (ilevkivskyi) [inline mypy/checker.py]: This should be `lvalue_attribute_type` (in your notations) here and one line above. I understand you are trying to duplicate the logic in `check_simple_assignment()`. Maybe instead modify the latter to accept an optional extra message?

File: mypy/checker.py [hunk 12]
```diff
@@ -2674,6 +2684,52 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
...
+                                    i.type.bases for i in lvalue_type.items))
+            else:
+                direct_lvalue_vars = lvalue_type.type.names
+                lvalue_bases = lvalue_type.type.bases
+                if '__metaclass__' in lvalue_type.type.names:
+                    # if the assigned metaclass is not class itself we have other problems
+                    metacls = lvalue_type.type.names['__metaclass__'].type.ret_type  # type: ignore
+                    lvalue_bases.append(metacls)
+                    attribute_is_defined = (attribute_is_defined or
+                                            metacls.has_readable_member(context.name))
```
Comment (ilevkivskyi) [inline mypy/checker.py]: Using the same message (`base class "..." defined ... as ...`) for both cases (base class and metaclass) is not good.

File: mypy/checker.py (hunk 10 above)
Comment (ilevkivskyi) [reply mypy/checker.py]: On a second thought, it seems to me, we may not show this additional info if descriptor protocol is involved, it may only add confusion, because the definition a user will see in the base class, is the descriptor type. But this is not a strong opinion.

File: mypy/checker.py [hunk 13]
```diff
@@ -2612,31 +2612,39 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
...
-            self.msg, context, chk=self)
+            self.msg, lvalue, chk=self)
         if not attribute_type.type.has_readable_member('__set__'):
             # If there is no __set__, we type-check that the assigned value matches
             # the return type of __get__. This doesn't match the python semantics,
             # (which allow you to override the descriptor with any value), but preserves
             # the type of accessing the attribute (even after the override).
-            rvalue_type = self.check_simple_assignment(get_type, rvalue, context)
+            rvalue_type = self.check_parent_member_assignment(instance_type, get_type,
+                                                              lvalue, attribute_type, rvalue)
```
Comment (ilevkivskyi) [inline mypy/checker.py]: Do you have a test case for a situation where `__get__` is involved?

File: mypy/checker.py [hunk 14]
```diff
@@ -2645,18 +2653,18 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
         dunder_set_type = expand_type_by_instance(bound_method, typ)
 
         # Here we just infer the type, the result should be type-checked like a normal assignment.
-        # For this we use the rvalue as type context.
+        # For this we use the rvalue as type lvalue.
```
Comment (ilevkivskyi) [inline mypy/checker.py]: The mass replace context -> lvalue went wrong.

File: mypy/checker.py [hunk 15]
```diff
@@ -2670,10 +2678,53 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
...
         # what a user expects.
-        rvalue_type = self.check_simple_assignment(set_type, rvalue, context)
+        rvalue_type = self.check_simple_assignment(set_type, rvalue, lvalue)
         infer = is_subtype(rvalue_type, get_type) and is_subtype(get_type, set_type)
         return rvalue_type if infer else set_type, get_type, infer
 
+    def check_parent_member_assignment(self, lvalue_base_type: Type,
+                                       lvalue_type: Type, lvalue: MemberExpr,
+                                       rvalue_type: Type, rvalue: Expression) -> Type:
+        """Type inherited member assignment.
```
Comment (ilevkivskyi) [inline mypy/checker.py]: I can't parse this as English sentence, could you please reformulate?

File: mypy/checker.py [hunk 16]
```diff
@@ -2670,10 +2678,53 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
...
+
+        This defers to check_simple_assignment if the attribute is local to the instance_type,
+        otherwise it will inform the user where the attribute type was originally defined.
```
Comment (ilevkivskyi) [inline mypy/checker.py]: I would say explicitly that it does exactly the same as `check_simple_assignment()`, but adds better error message if possible.

File: mypy/checker.py [hunk 17]
```diff
@@ -2670,10 +2678,53 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
...
+
+        Return the inferred rvalue_type.
+        """
+        if (isinstance(lvalue_base_type, Instance) and
+                lvalue.name not in lvalue_base_type.type.names):
```
Comment (ilevkivskyi) [inline mypy/checker.py]: Could you invert this check and put the last return here? This will save one indent level for the whole function body.

File: mypy/checker.py [hunk 18]
```diff
@@ -2670,10 +2678,53 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
...
+            metaclass = lvalue_base_type.type.metaclass_type
+            if metaclass and metaclass.has_readable_member(lvalue.name):
+                if lvalue.name in metaclass.type.names:
+                    meta_attribute = metaclass.type.names[lvalue.name].node
+                    if not (isinstance(meta_attribute, Var) and meta_attribute.info.is_generic()):
+                        return self.check_simple_assignment(
+                            lvalue_type, rvalue, lvalue,
+                            msg=message_registry.INCOMPATIBLE_TYPES_IN_ASSIGNMENT,
+                            lvalue_name=message_registry.METACLASS_MISMATCH.format(
+                                metaclass.type.defn.name))
```
Comment (ilevkivskyi) [inline mypy/checker.py]: This whole part is duplicated below. Can you try factoring it out?

File: mypy/checker.py [hunk 19]
```diff
@@ -2670,10 +2678,53 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
...
+                else:
+                    # attribute is defined further up in the metaclass MRO
+                    lvalue_warn = 'base class "{{}}" of {}'.format(
+                        message_registry.METACLASS_MISMATCH.format(
+                            metaclass.type.defn.name))
+                    lvalue_base_type = metaclass
+            if lvalue_base_type.has_readable_member(lvalue.name):
```
Comment (ilevkivskyi) [inline mypy/checker.py]: Is this just a performance optimization? If yes, I would remove it. It will not give much speed-up and the logic in this method is already hard to follow.

File: test-data/unit/check-functions.test [hunk 20]
```diff
@@ -2542,3 +2542,108 @@ def f() -> int: ...
...
+a = A(True)
+a().x = 9
+
+A.x = "" # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+A().x = "" # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+A(False).x = "" # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+A()().x = "" # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+A(False)().x = "" # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+A()()().x = "" # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+A(False)()().x = "" # E: Incompatible types in assignment (expression has type "str", variable has type "int")
```
Comment (ilevkivskyi) [inline test-data/unit/check-functions.test]: Why do you need al these test cases? What do they test?

File: test-data/unit/check-functions.test [hunk 21]
```diff
@@ -2542,3 +2542,108 @@ def f() -> int: ...
...
+A.b = True
+A(True)
+A(True).b = False
+A(True)(False)
+A(True)(False)(True).b = False
+A(True)(None)
+A(True)().b = False
+A(True)(None).b = True
+A(True)(None).b = 1 # E: Incompatible types in assignment (expression has type "int", variable has type "bool")
+A(True)()(False).b = "" # E: Incompatible types in assignment (expression has type "str", variable has type "bool")
```
Comment (ilevkivskyi) [inline test-data/unit/check-functions.test]: Same question here.

File: test-data/unit/check-functions.test [hunk 22]
```diff
@@ -2542,3 +2542,108 @@ def f() -> int: ...
...
+f().x = 0 # E: Incompatible types in assignment (expression has type "int", base class "A" defined the type as "str")
+f()().x = "baz"
+f()().x = 1 # E: Incompatible types in assignment (expression has type "int", base class "A" defined the type as "str")
+g().x = ""
+g().x = 2 # E: Incompatible types in assignment (expression has type "int", base class "A" defined the type as "str")
+
+[case testRecursiveCallable]
+from typing import Optional
+class A:
+    x = 0 # Type: int
```
Comment (ilevkivskyi) [inline test-data/unit/check-functions.test]: * You don't need to use type comments in tests.
* This is not a type comment because of a typo (uppercase `T`)
* It is not needed here

**Comment by hauntsaninja** [comment]
Closing this PR, since it's gotten stale and there are a number of merge conflicts. But you're most welcome to pick it back up :-)
